flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/ExternallyInducedSourceOperatorStreamTaskTest.java

```diff
@@ -0,0 +1,238 @@
+                        CheckpointOptions.alignedWithTimeout(CHECKPOINT_LOCATION, 123L),
+                        CheckpointOptions.unaligned(CHECKPOINT_LOCATION),
+                        CheckpointOptions.notExactlyOnce(
+                                CheckpointType.CHECKPOINT, CHECKPOINT_LOCATION))
+                .flatMap(
+                        options ->
+                                Stream.of(
+                                        new Object[] {options, true},
+                                        new Object[] {options, false}))
+                .collect(Collectors.toList());
+    }
+
+    @Test
+    public void testExternallyInducedSource() throws Exception {
+        final int numEventsBeforeCheckpoint = 10;
+        final int totalNumEvents = 20;
+        TestingExternallyInducedSourceReader testingReader =
+                new TestingExternallyInducedSourceReader(numEventsBeforeCheckpoint, totalNumEvents);
+        try (StreamTaskMailboxTestHarness<Integer> testHarness =
+                createTestHarness(new TestingExternallyInducedSource(testingReader), 0, null)) {
+            TestingExternallyInducedSourceReader runtimeTestingReader =
+                    (TestingExternallyInducedSourceReader)
+                            ((SourceOperator) testHarness.getStreamTask().mainOperator)
+                                    .getSourceReader();
+
+            CheckpointMetaData checkpointMetaData =
+                    new CheckpointMetaData(TestingExternallyInducedSourceReader.CHECKPOINT_ID, 2);
+            if (rpcFirst) {
+                testHarness.streamTask.triggerCheckpointAsync(
+                        checkpointMetaData, checkpointOptions);
+                testHarness.processAll();
+            } else {
+                do {
+                    testHarness.processSingleStep();
+                } while (!runtimeTestingReader.shouldTriggerCheckpoint().isPresent());
+                // stream task should block when trigger received but no RPC
+                assertFalse(testHarness.streamTask.inputProcessor.isAvailable());
+                CompletableFuture<Boolean> triggerCheckpointAsync =
+                        testHarness.streamTask.triggerCheckpointAsync(
+                                checkpointMetaData, checkpointOptions);
+                // process mails until checkpoint has been processed
+                while (!triggerCheckpointAsync.isDone()) {
+                    testHarness.processSingleStep();
+                }
+                // stream task should be unblocked now
+                assertTrue(testHarness.streamTask.inputProcessor.isAvailable());
+                testHarness.processAll();
+            }
+
+            int expectedEvents =
+                    checkpointOptions.getCheckpointType().isSavepoint()
+                                    && checkpointOptions.getCheckpointType().isSynchronous()
+                                    && checkpointOptions.getCheckpointType().shouldDrain()
```

Comment:
I needed to add this line or else the test failed for SAVEPOINT_SUSPEND compared to 1.15. In 1.15, we only receive 10 events, here we receive 20.

Reply:
That's fine. We unified the behaviour (stopping source before triggering a savepoint) only in 1.15. In 1.14 we still process records while triggering and cancel the job after the savepoint is taken.